### PR TITLE
Include taxonomic lineages in BIOM exports

### PR DESCRIPTION
**Pull request by audy**
## Status

- [x] **Ready**

## Description

- Include canonical taxonomic lineage in BIOM export

From One Codex to BIOM file:

```python
#!/usr/bin/env python3

import onecodex
import json

ocx = onecodex.Api()

# PRJNA692319: Metagenomic research in tundra Soils in Maritime Antarctica
proj = ocx.Projects.get("5e3ca50f0a8c4354")

samples = ocx.Samples.where(project=proj)

print(f"fetched {len(samples)} samples")

assert samples.taxonomy is not None

with open("tundra.biom", "w") as handle:
    json.dump(fp=handle, obj=samples.to_otu())

```

Load & plot BIOM file in Phyloseq

```R
#!/usr/bin/env Rscript

library(phyloseq)
library(magrittr)

phy <- import_biom('tundra.biom')

phy %>%
  transform_sample_counts(function(x) x/sum(x)) %>%
  prune_taxa(taxa_sums(.) > 0.03, .) %>%
  tax_glom('Rank7') %>%
  plot_bar(fill='Rank7')
```
Screenshot 2024-05-03 at 3 21 58 PM

## Related PRs
- [x] This PR is independent

## TODOs

- [x] Tests

### CODE REVIEW

File: onecodex/models/collection.py [hunk 1]
```diff
@@ -565,14 +576,19 @@ def _functional_results(self, **kwargs):
             self._cached[feature_map_key] = feature_map
         return self._cached[result_key], self._cached[feature_map_key]
 
-    def to_otu(self, biom_id=None):
-        """Transform a list of Classifications objects into a `dict` resembling an OTU table.
+    def to_otu(self, biom_id=None, include_ranks=CANONICAL_RANKS):
```
Comment (jairideout) [inline onecodex/models/collection.py:579]: `include_ranks` isn't being used in the method

File: onecodex/models/collection.py [hunk 2]
```diff
@@ -630,15 +654,21 @@ def to_otu(self, biom_id=None):
...
         otu["data"] = []
 
-        for present_taxa in sorted(rows):
-            # add the row entry
+        for tax_id in sorted(rows):
             row_id = len(otu["rows"])
-            otu["rows"].append(
-                {"id": present_taxa, "metadata": {"taxonomy": tax_ids_to_names[present_taxa]}}
-            )
+            tax_node = root.find(tax_id)
```
Comment (jairideout) [inline onecodex/models/collection.py:659]: it'd be nice to not call `root.find()` twice for each `tax_id`

File: onecodex/models/collection.py [hunk 3]
```diff
@@ -630,15 +654,21 @@ def to_otu(self, biom_id=None):
...
+            lineage = ([tax_node] + tax_node.ancestors())[::-1]
```
Comment (jairideout) [inline onecodex/models/collection.py:660]: nit: does this work?

```python
lineage = tax_node.ancestors() + [tax_node]
```

**Review by jairideout — CHANGES_REQUESTED** [review]
the only blocker is updating the method to use `include_ranks`

File: onecodex/models/collection.py (hunk 2 above)
Comment (audy) [reply onecodex/models/collection.py:659]: skbio is caching subsequent calls to `find` so this should be just as fast as keeping a separate map of `tax_id -> node`

**Pull request merged**